cobalt/network/disk_cache/cobalt_disk_cache.cc

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/network/disk_cache/cobalt_disk_cache.h"
+
+#include <string>
+#include <utility>
+
+#include "base/files/file_path.h"
+#include "base/metrics/field_trial.h"
+#include "base/task/task_scheduler/task_scheduler.h"
+#include "cobalt/network/disk_cache/cobalt_backend_impl.h"
+#include "net/base/cache_type.h"
+#include "net/base/net_errors.h"
+#include "net/disk_cache/backend_cleanup_tracker.h"
+#include "net/disk_cache/cache_util.h"
+#include "net/disk_cache/disk_cache.h"
+#include "net/disk_cache/memory/mem_backend_impl.h"
+
+namespace {
+
+// Builds an instance of the backend depending on platform, type, experiments
+// etc. Takes care of the retry state. This object will self-destroy when
+// finished.
+class CacheCreator {
+ public:
+  CacheCreator(const base::FilePath& path, int64_t max_bytes,
+               net::NetLog* net_log,
+               std::unique_ptr<::disk_cache::Backend>* backend,
+               net::CompletionOnceCallback callback,
+               cobalt::network::URLRequestContext* url_request_context);
+
+  net::Error TryCreateCleanupTrackerAndRun();
+
+  // Creates the backend, the cleanup context for it having been already
+  // established... or purposefully left as null.
+  net::Error Run();
+
+ private:
+  ~CacheCreator();
+
+  void DoCallback(int result);
+
+  void OnIOComplete(int result);
+
+  const base::FilePath path_;
+  bool retry_;
+  int64_t max_bytes_;
+  std::unique_ptr<::disk_cache::Backend>* backend_;
+  net::CompletionOnceCallback callback_;
+  std::unique_ptr<::disk_cache::Backend> created_cache_;
+  net::NetLog* net_log_;
+  scoped_refptr<::disk_cache::BackendCleanupTracker> cleanup_tracker_;
+  cobalt::network::URLRequestContext* url_request_context_;
+
+  DISALLOW_COPY_AND_ASSIGN(CacheCreator);
```

Comment:
Just a note, not necessarily needed to fix here, because this is a cleanup task that we will have to start doing soon (b/315500709): We need to switch from using these macros to `= delete` syntax for the corresponding constructors.